### DISCUSSION TITLE
Make GraphQL schema cacheable

### DISCUSSION THREAD

**Discussion by andris-sevcenko** [craftcms/cms#8968] [Ideas]
@brandonkelly - So this is interesting - we have a headless site paired with a static site generator, and whenever the site regenerates we are seeing the server spend almost all of its time rebuilding the schema via `Gql::getSchemaDef` - here is some profiling we did on New Relic to see where we were spending so much time - as you can see, a ton of time is being spent in PHP, and actually very little in MySQL:

<img width="937" alt="Screen Shot 2020-12-03 at 10 35 17 PM" src="https://user-images.githubusercontent.com/273266/101118808-dabc6880-35b7-11eb-8bcc-37d384e2b8a5.png">

We have our rebuild pretty well-optimized (GQL caching as well as GQL batching, on pages with multiple calls), but digging into the API GQL controller I can see that even with GQL caching, it does quite a bit of work pulling the schema definition on every request. Would it be possible for the schema to be cached once the query has been validated? Seems like there is a lot of time being spent here when it probably doesn't have to.

_Originally posted by @mildlygeeky in https://github.com/craftcms/cms/issues/6639#issuecomment-738549165_

**Comment by andris-sevcenko** [comment]
@mildlygeeky yeah, the schema definition is required for every call, otherwise, it's impossible to validate the incoming query against it.

We've tried to cache the resulting schema, but, for now, it's impossible, as GraphQL fields are defined within Closures, to help to resolve circular dependencies. When you try to serialize anything with a Closure in it, PHP throws a hissy fit about it, so there's no way _currently_ to cache generated schema.

For Craft 4.0, I hope to add something like https://github.com/opis/closure to help alleviate this, but that's not going to happen for 3.6.

**Comment by mildlygeeky** [comment]
Aaah - that makes sense around closures being the reason why you can't simply memoize/serialize. Thanks for the explanation, @andris-sevcenko, and hope to see this in a future release.

**Comment by mildlygeeky** [comment]
@brandonkelly @andris-sevcenko if I could put together a PR for this, might this something that could get into 3.x (3.7)? Really seeing a lot of stress on Craft for sites using heavy use of GQL, and would be happy to look at caching this (thought is it would respect Craft's overall GQL caching, and break the cache for this in the same way).

**Comment by andris-sevcenko** [comment]
@mildlygeeky, I decided to give it a shot and see what I can come up with.

So, the problem here was that the Schema on the PHP end is a complex beast. It's not only a GraphQL schema to validate a query against, but it also holds all the knowledge on how to resolve different fields and whatnot. Further, since entries can relate entries, for example, the fields _have_ to be defined as callbacks, otherwise, you end up with infinite loops. Now, to go to an even more fun place, callbacks/closures cannot be serialized, so it's hard to cache without any wizardry.

Thankfully, there are [excellent libraries out there](https://opis.io/closure/3.x/serialize.html#serialize-unserialize-arbitrary-objects) that can do this for us, and we should assume that it's with decent performance. Maybe not perfect, but probably not _too_ much overhead on top of PHPs serialization goes into wrapping all the closures in classes so they can be serialized.

**Long story short, I made that happen.**

Now, let's take a detour. If you're not in the mood for detours and rambles, skip to the end.

I don't recall if you were a fan of Game Of Thrones, but we can probably all agree that George RR Martin is a master of plot twists and great stories in general. Did you know he started by writing science fiction stories?

I was not aware of this until I went hunting for the science fiction story I recalled that pairs up nicely with this GH issue - it was a story I read a long time ago as part of [this book](https://www.goodreads.com/book/show/666584.100_Great_Science_Fiction_Short_Short_Stories). The story, it turns out, was by George RR Martin, and it was called FTA. If you're not into science fiction, you can read the plot [synopsis here](https://scifi.stackexchange.com/questions/84453/scientist-discovers-hyperspace-with-a-twist/84454). If you are, well, that book has some good stories.

Anyway, as you might have guessed, caching is way slower than generating the Schema every time. About 3x slower.

I'll look into some more, uhh, "experimental" solutions that might be available on some hosts.

**Comment by andris-sevcenko** [comment]
In particular, maybe [preloading](https://www.php.net/manual/en/opcache.preloading.php) can be made to work here, but that requires PHP 7.4, so the Craft 4.0 milestone still remains.

**Comment by andris-sevcenko** [comment]
Brainstormed a little bit with @Wiejeben over Discord about this.

Currently, seems like the best (maybe the only) realistic approach here is to leverage [Swoole](https://www.swoole.co.uk/docs/) here as a dedicated GraphQL server which holds the Schemas in the memory.

Then, using the `Gql::EVENT_BEFORE_EXECUTE_GQL_QUERY` a plugin could just forward the request to the Swoole server, which would execute the query then and return the results. Obviosuly, this is a rough draft of how that might work, but it's something.

**Comment by mildlygeeky** [comment]
> **Long story short, I made that happen.**
> 
> Now, let's take a detour. If you're not in the mood for detours and rambles, skip to the end.
> 
> I don't recall if you were a fan of Game Of Thrones, but we can probably all agree that George RR Martin is a master of plot twists and great stories in general. Did you know he started by writing science fiction stories?
> 
> I was not aware of this until I went hunting for the science fiction story I recalled that pairs up nicely with this GH issue - it was a story I read a long time ago as part of [this book](https://www.goodreads.com/book/show/666584.100_Great_Science_Fiction_Short_Short_Stories). The story, it turns out, was by George RR Martin, and it was called FTA. If you're not into science fiction, you can read the plot [synopsis here](https://scifi.stackexchange.com/questions/84453/scientist-discovers-hyperspace-with-a-twist/84454). If you are, well, that book has some good stories.
> 
> Anyway, as you might have guessed, caching is way slower than generating the Schema every time. About 3x slower.
> 
> I'll look into some more, uhh, "experimental" solutions that might be available on some hosts.

Friends, this is why I will forever "stan" for @andris-sevcenko. You crack me up, man.

**Comment by mildlygeeky** [comment]
One way to support this without fully building Swoole support into core may be to include a gql.php script in web, and mount that as an endpoint for the craftcms/docker containers with Swoole's official docker image (https://github.com/swoole/docker-swoole/tree/master/examples/03-nginx) - we're going to try this out for a new build, and will report back on how it goes.

**Reply by jamesedmonston** [reply]
@mildlygeeky did you have any luck with this? We're running into issues with a site that has a particularly large schema (the JSON for it is 17MB) – uncached queries are taking between 2-3 seconds to resolve.

**Reply by mildlygeeky** [reply]
I haven't - ran into many issues with Swoole.

At this point I'm considering using CloudFlare workers to cache the response for non-mutate GQL calls (and then having Craft clear the cache when content or the schema changes): https://developers.cloudflare.com/workers/examples/cache-post-request

**Reply by andris-sevcenko** [reply]
I can confirm that there are quite a few issues with Swoole to make it stable and reliable. For what it's worth, I was able to accomplish this with minor changes to Craft itself.

There are still quite a few things to take care of to make it even alpha-worthy, but it might happen for Craft 4 with a plugin in compatible environments.

https://user-images.githubusercontent.com/1891118/134677738-1fe7fc53-1209-4036-81a2-7f007e1fb557.mov

:

**Reply by mildlygeeky** [reply]
Oh, now you're just teasing me @andris-sevcenko

**Reply by andris-sevcenko** [reply]
Error handling is non-existent and when Swoole goes down, it goes down hard. It's not ready for commitment yet, @mildlygeeky  :)

**Reply by Wiejeben** [reply]
I feel like this would best be implemented as some kind of daemon which you can configure Craft to forward GQL queries to, whenever the daemon is down it could fallback to the current way of handling GQL queries. This would also add a layer of security to not expose the Swoole server on whatever alternative port (plus how to deal with HTTPS and all that).

**Reply by andris-sevcenko** [reply]
@Wiejeben Swoole can run as a daemon, too. Also, in this case, it's configured to listen to the 127.0.0.1 network address, so is already inaccessible from outside.

Or am I misunderstanding you, @Wiejeben?

**Comment by FreekVR** [comment]
Not sure if this helps, but lighthouse-php, the Laravel package for building GraphQL schemas, does have caching. They seem to have solved this by writing the compiled schema to the filesystem as a PHP file;

https://lighthouse-php.com/5/performance/schema-caching.html
https://github.com/nuwave/lighthouse/blob/16de64b257589b3901c4aedb15920625c605caee/src/Schema/AST/ASTBuilder.php#L97

Might this be a route that could be taken for Craft, too?

**Reply by brandonkelly** [reply]
Ohh interesting. Thanks for pointing that out @FreekVR

**Reply by andris-sevcenko** [reply]
That is not that different, in the end. There's quite a bit of overhead to make the AST cacheable and you still end up with serializing and unserializing a bunch of stuff. Also, there's additional overhead to detaching and re-attaching all the lambda functions (resolvers and field definitions at the very least).

For what it's worth, it's not an answer I've come up with out of the blue - "it's more efficient to efficiently lazy-load than cache" is a sentiment that's also shared by @markhuot (author of the original CraftQL plugin) and @vladar (author/contributor of `webonyx/graphql-php`)

**Comment by Wiejeben** [comment]
Unless this is now something on the roadmap of Craft, I feel like this question has actually gone unanswered.

**Reply by mildlygeeky** [reply]
Agreed - GraphQL performance is real tough on a large schema as-is. Right now we’re using Stellate (formerly GraphCDN) to get around it, but would really like to see faster GQL out of the origin endpoint, especially for everything we can’t cache.

**Comment by tekstrand** [comment]
We're seeing this issue as well under peak loads, is this still being considered for improvement?

**Reply by brandonkelly** [reply]
Still on our radar, but the initial ideas we’ve explored haven’t really moved the needle.

**Comment by dwgebler** [comment]
@andris-sevcenko did you ever manage to get anything working for caching the schema definition via `opis/closure`? I've been trying to do this as a workaround and although I can produce a serialized representation of the schemadef, when I try to unserialize it I get an error ("error at offset..." PHP error which indicates a corrupted serialized string)
